services/src/datasets/external/netcdfcf/database.rs

```diff
@@ -0,0 +1,1438 @@
+        NetCdfDatabaseListingConfig {
+            provider_id,
+            collection,
+            file_name,
+            group_path,
+            options,
+        }: NetCdfDatabaseListingConfig<'_>,
+        id_fn: Arc<ID>,
+    ) -> Result<Option<LayerCollection>>;
+
+    async fn layer(
+        &self,
+        provider_id: DataProviderId,
+        layer_id: &LayerId,
+        file_name: &str,
+        group_path: &[String],
+        entity: usize,
+    ) -> Result<Option<Layer>>;
+
+    async fn loading_info(
+        &self,
+        provider_id: DataProviderId,
+        file_name: &str,
+        group_path: &[String],
+        entity: usize,
+    ) -> Result<Option<GdalMetaDataList>>;
+
+    #[cfg(test)]
+    /// only for testing
+    async fn test_execute_with_transaction<T>(
+        &self,
+        statement: &T,
+        params: &[&(dyn tokio_postgres::types::ToSql + Sync)],
+    ) -> Result<(), tokio_postgres::Error>
+    where
+        T: ?Sized + tokio_postgres::ToStatement + Sync;
+}
+
+/// Gets a connection from the pool
+macro_rules! get_connection {
+    ($connection:expr) => {
+        $connection
+            .get()
+            .await
+            .boxed_context(error::DatabaseConnection)?
+    };
+}
+
+/// Starts a read-only snapshot-isolated transaction for a multiple read
+macro_rules! readonly_transaction {
+    ($connection:expr) => {
+        $connection
+            .build_transaction()
+            .read_only(true)
+            .deferrable(true) // get snapshot isolation
+            .start()
+            .await
+            .boxed_context(error::DatabaseTransaction)?
+    };
+}
+
+/// Starts a deferred write transaction
+macro_rules! deferred_write_transaction {
+    ($connection:expr) => {{
+        let transaction = $connection
+            .transaction()
+            .await
+            .boxed_context(error::DatabaseTransaction)?;
+
+        // check constraints at the end to speed up insertions
+        transaction
+            .batch_execute("SET CONSTRAINTS ALL DEFERRED")
+            .await
+            .boxed_context(error::UnexpectedExecution)?;
+
+        transaction
+    }};
+}
+
+#[async_trait]
+impl<Tls> NetCdfCfProviderDb for PostgresDb<Tls>
+where
+    Tls: MakeTlsConnect<Socket> + Clone + Send + Sync + 'static + std::fmt::Debug,
+    <Tls as MakeTlsConnect<Socket>>::Stream: Send + Sync,
+    <Tls as MakeTlsConnect<Socket>>::TlsConnect: Send,
+    <<Tls as MakeTlsConnect<Socket>>::TlsConnect as TlsConnect<Socket>>::Future: Send,
+{
+    async fn lock_overview(&self, provider_id: DataProviderId, file_name: &str) -> Result<bool> {
+        let connection = get_connection!(self.conn_pool);
+
+        let updated_rows = connection
+            .execute(
+                "
+                INSERT INTO ebv_provider_locks (
+                    provider_id,
+                    file_name
+                ) VALUES (
+                    $1,
+                    $2
+                ) ON CONFLICT DO NOTHING;
```

Comment:
shouldn't it be an error if we want to lock something that's already locked?

Reply:
It outputs true or false if we were able to lock it.